### ISSUE TITLE
Document the correct way to deprecate a function in Twisted

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5645]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5645|
|Type|enhancement|
|Created|2012-04-13 20:43:09Z|
|Branch|https://github.com/twisted/twisted/tree/deprecation-note-list-5645|

We have a pretty strict process for deprecations.  It's simple, but the steps all need to be followed exactly:

  * Do not change the function's behavior as part of this process
  * Cause usage of the function to emit a `DeprecationWarning`
    * either call `warnings.warn`
    * or use one of the APIs in `twisted.python.deprecate`
  * The warning text must include the version of Twisted in which the function is first deprecated (which will always be a version in the future)
  * The warning text should recommend a replacement, if one exists
  * The warning must "point to" the code which called the function.  For example, in the normal case, this means `stacklevel=2` passed to `warnings.warn`.
  * There must be a unit test which verifies these things.
    * Unit tests can use `TestCase.flushWarnings` or `callDeprecated` or one of the other helpers.
  * A `.removal` news fragment must be added.

Maybe I forgot some things, too.

<details><summary>Searchable metadata</summary>

```
trac-id__5645 5645
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_deprecation_note_list_5645 branches/deprecation-note-list-5645
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1334349789000000 1334349789000000
changetime__1462275964214598 1462275964214598
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_1"></a>|@tomprince commented|
|-|-|

  * Tests of the deprecated function must be left, and appropriate `.suppress` attribute added to avoid emitting the `DeprecationWarning` for them.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_2"></a>|@tomprince commented|
|-|-|

[#6266](https://github.com/twisted/twisted/issues/6266) is relevant for making remaining tests not emit deprecation warnings.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince commented|
|-|-|

[#3266](https://github.com/twisted/twisted/issues/3266) is also related.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

(In [47329]) Branching to deprecation-note-list-5645.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

I have updated the documentation to include the list from the ticket description.

Please check that it make sense and feel free to commit your suggestions

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @adiroiban|
|-|-|

We should probably improve this a bit more, but the doc builder is happy and this is a good first step towards a more expansive to-do guide.  Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set status to `closed`|
|-|-|

(In [47340]) Merge deprecation-note-list-5645: To-do guide for deprecating.

Author: adiroiban
Reviewer: glyph
Fixes: [#5645](https://github.com/twisted/twisted/issues/5645)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

Thanks for the review.

I assume that when a new developer will try to read the documentation, she can provide valuable feedback and update the list.

I am already a bit familiar with the deprecation process and I find it hard to know what parts are not clear for new developers.
